homework/tests/example.spec.js

```diff
@@ -1,31 +1,4015 @@
 // @ts-check
 const { test, expect } = require('@playwright/test');
 
+test.beforeEach(async ({ page }) => {
+  await page.goto('https://testsheepnz.github.io/BasicCalculator');
+});
+
 const data = [
   'Prototype',
-  '1',
-  '2',
-  '3',
-  '4',
-  '5',
-  '6',
-  '7',
-  '8',
-  '9'
+  //'1',
+  //'2',
+  //'3',
+  //'4',
+  //'5',
+  //'6',
+  //'7',
+  //'8',
+  //'9'
 ]
 
+const ops = [
+  'Add',
+  'Subtract',
+  'Multiply',
+  'Divide',
+  'Concatenate'
+]
 
```

Comment:
It is good that you thought of so many different test-cases to test this calculator application, but you could have covered  only the main functionality which you described in "ops" constant above (5 tests) and include some basic functionality checks like if fields are present and if they are enabled in 1 test, if invalid field error is present in another test, and if "clear" button works as expected in another one. In total we would have 8 tests which would run on every single possible build. Now we have many repetitive test-cases which brings so little value, e.g.
1. 'Should be able to input negative float as -11.2 second number'
2. 'Should be able to input negative float as -11.2 second number'

Try to create a function which checks if field is visible and if field is enabled (that would indicate that user can write in the input) which could be re-used. This is not finished code snippet which I hope will help you to figure this out:

```
  async assertFieldExistsAndIsEnabled(fieldId) {
    const visible = await this.page.isVisible(fieldId);
    assert

    const enabled = await this.page.isEnabled(fieldId);
    assert
  }
```

    test(`Fields exists and are enabled`, async () => {
      await assertFieldExistsAndIsEnabled("#number1Field");
       ...
       ...
    });